src/common/include/common/config.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2020 Evan Lojewski
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#ifndef _COMMON_CONFIG_H
+#define _COMMON_CONFIG_H
+
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+struct config_t;
+typedef struct config_t config_t;
+
+typedef void (*config_callback_t)(struct config_t *entry);
+
+typedef struct {
+    int32_t max_value;
+    int32_t min_value;
+    int32_t value;
+} config_value_t;
+
+struct config_t {
+    const char *config_short;
+    const char *config_desc;
```

Comment:
Would you please elaborate on why having strings concerns you? 

At a minimum, I think the human readable name (config_short) should be here. I don't see a downside for having both really - there's currently an extra 96KB of flash that isn't in use. The string copy code is also limited to the buffer size, 256 bytes, and so there is not really a risk of overrunning memory. That said, the config save code is not limited in the same way - it's only limited to the side of the config space. I can go ahead and limit this further, however.

Personally, I'm a lot more concerned about the ability to erase and write to flash. To me it seems more likely that the code will accidentally brick itself. One option to make this safer is to explicitly check bounds in the flash application.

The reason I did it this way is to enable easy autoconfig and so multiple tools don't need to be kept in sync. Possible host-side tools include the ectool, possibly a settings plugin for gnome, or a UEFI application. With the way the code is now, valid settings can be read from the firmware and customized automatically. If the name or description were moved, you now need to both update the host tool and the firmware at the same time. For a gnome setting plugin, this isn't really supported right now. If the ectool is automatically installed in Pop!, Then this also wouldn't be ensured to be in sync. for a UEFI application, you could enforce this, howerver.

One other benifit is that it's very easy to find the setting in a config dump when manually inspecting.

All that said, moving just the description would still allow for the human readable name (just no description).

Reply:
> Would you please elaborate on why having strings concerns you?

I'm mainly concerned with string lengths: the space they'll take up and reading from/writing to flash. I was also thinking about changing these strings (requiring a firmware update) and translating these strings (but I suppose gettext would work for user-space applications?).

For example, the setting could be represented by a single byte. This would mean only reading/writing 1 byte instead of _n_ bytes for the identifier. Each entry would then also be the same size. Conceptually, this is much simpler to iterate and find settings (no need for `flash_strncmp()` or tracking lengths).

(I can be convinced either way, though. I've only worked with firmware that has settings defined at compile time, so I don't have much exposure with this stuff. Even the Linux kernel embeds the name and description into modules.)

Reply:
I agree that the search isn't the most optimal, and it takes up more space. I guess it comes down to priorities. I tend to lean towards autoconfig whenever possible. (I also don't have variable length strings within the firmware I normally work on, although most of the firmware I work on is configurable by the user in various other ways).

You also have a good point about translations, I hadn't thought about that. You'd still need to know the strings (or an I'd) in there user app to have a translation.

I could be convinced to have, say, a 4 byte id such as 'KFNL' for keyboard function lock and 'CCDE' for camera enable, for instance. (Maybe even just 16bits is enough). This could either replace or supliment the human readable name/desc while still having some meaning.

The short Id (1 to 4 bytes) would be used for, say, searches and storing in flash.

I can update everything in this pull request to use the smaller 1-4 byte id for searches / commands. I propose leaving the desc / name so that at least English translations can be discovered, but if you'd prefer I can (reluctantly) move those to the ectool.

If you think it's worth doing, as a second step, I could also see adding a resource table in flash (longer descriptions, translations, keymaps, other could be stored here). This could then completely replace the larger strings.
The resource table could be stored in end of flash (outside of the 64kb region) and read by a command. 

I can add another pull request adding a resource table. I'd then add an command to allow the host to read out resource, and move the English translations to said table. As a longer term feature, it might be nice to allow keymaps to be set here, not just read. A user could generate them using something like the qmk configurator.

Reply:
> The short Id (1 to 4 bytes) would be used for, say, searches and storing in flash.

That's a good compromise. Keep the strings, and add an identifier for enumeration/lookup. I'd be fine with either 1 byte (binary values) or 4 bytes (ASCII short-code). 2 bytes would be too much for potential config identifiers\* and too short for an ASCII identifier.

\* Unless a change in semantics for one results in a new identifier, similar to adding new commands while maintaining compatibility with old ones.

(Using a 2 byte value like that sounds like it covers my concerns about changing the format of a config in the future.)

I don't think the resource table is worth doing; Descriptions and translations can be handled by higher level tools and documentation (our BIOS firmware is also English-only). The keymap idea is intriguing, though, since everyone wants something slightly different.

Reply:
The code now uses a 4 byte ascii identifier.

Note that in SDCC setting a uint32_t value to  'CCDE'  actually only sets it to 'C', unlike for normal 32bit targeting compilers. As such I have it using {'C', 'C', 'D', 'E'} instead even though I'd prefer not to.